src/core/lib/security/credentials/tls/grpc_tls_crl_provider.h

```diff
@@ -77,6 +79,46 @@ class CertificateInfoImpl : public CertificateInfo {
   const std::string issuer_;
 };
 
+class DirectoryReloaderCrlProvider : public CrlProvider {
+ public:
+  static absl::StatusOr<std::shared_ptr<CrlProvider>>
+  CreateDirectoryReloaderProvider(
+      absl::string_view directory, std::chrono::seconds refresh_duration,
+      std::function<void(absl::Status)> reload_error_callback);
+  // ~DirectoryReloaderCrlProvider() override;
+};
+
+// Defining this here lets us hide implementation details (and includes) from
+// the header in include
+class DirectoryReloaderCrlProviderImpl
+    : public DirectoryReloaderCrlProvider,
+      public std::enable_shared_from_this<DirectoryReloaderCrlProviderImpl> {
+ public:
+  DirectoryReloaderCrlProviderImpl(::absl::string_view directory,
+                                   ::std::chrono::seconds duration,
+                                   ::std::function<void(absl::Status)> callback)
+      : crl_directory_(directory),
+        refresh_duration_(duration),
+        reload_error_callback_(callback),
+        event_engine_(
+            grpc_event_engine::experimental::GetDefaultEventEngine()) {}
+  ~DirectoryReloaderCrlProviderImpl() override;
+  std::shared_ptr<Crl> GetCrl(const CertificateInfo& certificate_info) override;
```

Comment:
Shouldn't `GetCrl` (and the constructor/destructor) be the only public method?

Reply:
This is in `src/core`, not in `include` so it's not exposed, this is all internal. But yes, it'll be cleaner for much of this to be private - I made the change. `ScheduleReload` and `Update` are called in the function `CreateDirectoryReloaderProvider` that is not a class member,  so they had to be public with the current design.